runtime/src/main/java/org/corfudb/runtime/view/QuorumReplicationView.java

```diff
@@ -0,0 +1,117 @@
+
+            LogData ld = new LogData(DataType.DATA, b);
+            ld.setBackpointerMap(backpointerMap);
+            ld.setStreams(stream);
+            ld.setGlobalAddress(address);
+
+            // FIXME
+            if (data instanceof LogEntry) {
+                ((LogEntry) data).setRuntime(getLayout().getRuntime());
+                ((LogEntry) data).setEntry(ld);
+            }
+
+            payloadBytes = b.readableBytes();
+            CompletableFuture<Boolean>[] futures = new CompletableFuture[numUnits];
+            for (int i = 0; i < numUnits; i++) {
+                log.trace("Write[{}]: quorum {}/{}", address, i + 1, numUnits);
+                // sending the message to all logging units and ignoring the exception.
+                futures[i] = getLayout().getLogUnitClient(address, i).write(address, stream, 0L, data, backpointerMap);
+            }
+            Boolean result = CFUtils.getUninterruptibly(QuorumFutureFactory.getQuorumFuture(futures));
+            log.trace("Write[{}]: {}", address, result);
+        }
+        return payloadBytes;
+    }
+
+    /**
+     * Read the given object from an address, using the replication method given.
+     *
+     * @param address The address to read from.
+     * @return The result of the read.
+     */
+    @Override
+    public LogData read(long address) {
+        int numUnits = getLayout().getSegmentLength(address);
+        log.trace("Read[{}]: chain {}/{}", address, numUnits, numUnits);
+        CompletableFuture<ReadResponse>[] futures = new CompletableFuture[numUnits];
+            for (int i=0; i<numUnits; i++) {
+                futures[i]=getLayout().getLogUnitClient(address, i).read(address);
+        }
+        ReadResponse readResponse = CFUtils.getUninterruptibly(QuorumFutureFactory.getQuorumFuture(futures));
+        Map<Long, LogData> rs = readResponse.getReadSet();
+        return rs.get(address);
+    }
+
+    /**
+     * Read a stream prefix, using the replication method given.
+     *
+     * @param stream the stream to read from.
+     * @return A map containing the results of the read.
+     */
+    @Override
+    public Map<Long, LogData> read(UUID stream, long offset, long size) {
+        // TODO: when quorum replication is used, scan
+        throw new UnsupportedOperationException("not supported in quorum replication");
+    }
+
+    /**
+     * Fill a hole at an address, using the replication method given.
+     *
+     * @param address The address to hole fill at.
+     */
+    @Override
+    public void fillHole(long address) throws OverwriteException {
+        int numUnits = getLayout().getSegmentLength(address);
+        CompletableFuture<Boolean>[] futures = new CompletableFuture[numUnits];
+        for (int i=0; i<numUnits; i++) {
+            log.trace("fillHole[{}]: quorum {}/{}", address, i + 1, numUnits);
+            futures[i]=getLayout().getLogUnitClient(address, i).fillHole(address);
```

Comment:
Yes, I am working on this. In order to make "fill hole" testable, I created a special rule (with your framework), that "damages" the sequencer and makes it return duplicate results. It even helped me to find one small mistake in the future above. Unfortunately the rule is still "red" and has some issues with the "chain replication" too. I believe I will be able to stabilize it and update my request until our meeting on Monday.